flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java

```diff
@@ -520,72 +439,128 @@ protected YarnClusterClient deployInternal() throws Exception {
 			LOG.warn("This YARN session requires " + totalMemoryRequired + "MB of memory in the cluster. "
-				+ "There are currently only " + freeClusterMem.totalFreeMemory + "MB available." + noteRsc);
+				+ "There are currently only " + freeClusterResources.totalFreeMemory + "MB available." + noteRsc);
 
 		}
-		if (taskManagerMemoryMb > freeClusterMem.containerLimit) {
+		if (taskManagerMemoryMb > freeClusterResources.containerLimit) {
 			LOG.warn("The requested amount of memory for the TaskManagers (" + taskManagerMemoryMb + "MB) is more than "
-				+ "the largest possible YARN container: " + freeClusterMem.containerLimit + noteRsc);
+				+ "the largest possible YARN container: " + freeClusterResources.containerLimit + noteRsc);
 		}
-		if (jobManagerMemoryMb > freeClusterMem.containerLimit) {
+		if (jobManagerMemoryMb > freeClusterResources.containerLimit) {
 			LOG.warn("The requested amount of memory for the JobManager (" + jobManagerMemoryMb + "MB) is more than "
-				+ "the largest possible YARN container: " + freeClusterMem.containerLimit + noteRsc);
+				+ "the largest possible YARN container: " + freeClusterResources.containerLimit + noteRsc);
 		}
 
 		// ----------------- check if the requested containers fit into the cluster.
 
-		int[] nmFree = Arrays.copyOf(freeClusterMem.nodeManagersFree, freeClusterMem.nodeManagersFree.length);
+		int[] nmFree = Arrays.copyOf(freeClusterResources.nodeManagersFree, freeClusterResources.nodeManagersFree.length);
 		// first, allocate the jobManager somewhere.
 		if (!allocateResource(nmFree, jobManagerMemoryMb)) {
 			LOG.warn("Unable to find a NodeManager that can fit the JobManager/Application master. " +
 				"The JobManager requires " + jobManagerMemoryMb + "MB. NodeManagers available: " +
-				Arrays.toString(freeClusterMem.nodeManagersFree) + noteRsc);
+				Arrays.toString(freeClusterResources.nodeManagersFree) + noteRsc);
 		}
 		// allocate TaskManagers
 		for (int i = 0; i < taskManagerCount; i++) {
 			if (!allocateResource(nmFree, taskManagerMemoryMb)) {
 				LOG.warn("There is not enough memory available in the YARN cluster. " +
 					"The TaskManager(s) require " + taskManagerMemoryMb + "MB each. " +
-					"NodeManagers available: " + Arrays.toString(freeClusterMem.nodeManagersFree) + "\n" +
+					"NodeManagers available: " + Arrays.toString(freeClusterResources.nodeManagersFree) + "\n" +
 					"After allocating the JobManager (" + jobManagerMemoryMb + "MB) and (" + i + "/" + taskManagerCount + ") TaskManagers, " +
 					"the following NodeManagers are available: " + Arrays.toString(nmFree)  + noteRsc);
 			}
 		}
 
-		ApplicationReport report = startAppMaster(null, yarnClient, yarnApplication);
+		return new ClusterSpecification(
+			jobManagerMemoryMb,
+			taskManagerMemoryMb,
+			clusterSpecification.getNumberTaskManagers(),
+			clusterSpecification.getSlotsPerTaskManager());
+
+	}
+
+	protected void logClusterSpecification(ClusterSpecification clusterSpecification) {
```

Comment:
I would remove this method as it is only used once and isn't really more readable than the inline alternative.